test/extended/cli/set-image.go

```diff
@@ -0,0 +1,128 @@
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(out).To(o.ContainSubstring("image-registry.openshift-image-registry.svc:5000/e2e-test-oc-set-image-"))
+		o.Expect(out).To(o.ContainSubstring("/ruby@sha256:"))
+
+		g.By("repeating basic image updates to ensure nothing changed")
+		err = oc.Run("set").Args("image", "dc/test-deployment-config", "ruby-helloworld=ruby:2.7-ubi8", "--source=istag").Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		out, err = oc.Run("get").Args("dc/test-deployment-config", "-o", "jsonpath='{.spec.template.spec.containers[0].image}'").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(out).To(o.ContainSubstring("image-registry.openshift-image-registry.svc:5000/e2e-test-oc-set-image-"))
+		o.Expect(out).To(o.ContainSubstring("/ruby@sha256:"))
+
+		g.By("testing invalid image tags")
+		err = oc.Run("set").Args("image", "dc/test-deployment-config", "ruby-helloworld=ruby:XYZ", "--source=istag").Execute()
+		o.Expect(err).To(o.HaveOccurred())
+
+		err = oc.Run("set").Args("image", "dc/test-deployment-config", "ruby-helloworld=ruby:XYZ", "--source=isimage").Execute()
+		o.Expect(err).To(o.HaveOccurred())
+
+		g.By("setting a different, valid image on deployment config")
+		err = oc.Run("set").Args("image", "dc/test-deployment-config", "ruby-helloworld=nginx").Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		out, err = oc.Run("get").Args("dc/test-deployment-config", "-o", "jsonpath='{.spec.template.spec.containers[0].image}'").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(out).To(o.ContainSubstring("nginx"))
+
+		g.By("setting a different, valid image on pod")
+		err = oc.Run("set").Args("image", "pod/hello-openshift", "hello-openshift=nginx").Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		out, err = oc.Run("get").Args("pod/hello-openshift", "-o", "jsonpath='{.spec.containers[0].image}'").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(out).To(o.ContainSubstring("nginx"))
+
+		g.By("setting a different, valid image tag on pod")
+		err = oc.Run("set").Args("image", "pod/hello-openshift", "hello-openshift=nginx:1.9.1").Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+
+		out, err = oc.Run("get").Args("pod/hello-openshift", "-o", "jsonpath='{.spec.containers[0].image}'").Output()
+		o.Expect(err).NotTo(o.HaveOccurred())
+		o.Expect(out).To(o.ContainSubstring("nginx:1.9.1"))
+
+		g.By("setting a different, valid image on multiple resources")
+		err = wait.Poll(time.Second, 2*time.Minute, func() (bool, error) {
+			// during the set image, it's possible that the temporary deployer pod existed, but the moment we invoke the patches the pod is gone, b/c it completed its task, so we retry several times
+			err := oc.Run("set").Args("image", "pods,dc", "*=ruby:2.7-ubi8", "--all", "--source=imagestreamtag").Execute()
+			return err == nil, nil
```

Comment:
Should we continue retry only for the error we care(pod is not found)?, or only error can be thrown is that?

Reply:
The only other error I can think of is a networking/cluster down issue, and if that happens the next commands will fail as well.

Reply:
My original suggestion was to ignore only a specific error `test-deployment-config-1-deploy not found` from https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_kubernetes/1243/pull-ci-openshift-kubernetes-master-e2e-agnostic-cmd/1516788704302075904 since the retries might still stumble on it.